### ISSUE TITLE
False-positives for RCS1213

### ISSUE THREAD

**Issue by jhinder** [JosefPihrt/Roslynator#234]
Referencing a field, method or property inside an indexer or a property that is not expression-bodied will result in a false-positive for RCS1213, i.e. it will suggest to remove the declaration despite it being referenced by other code.

Repro 1: RCS1213 suggests removing `field`.

```c#
public class RCS1213Repro_1
{
    private object field;
    public object Property
    {
        get => field;
        set
        {
            if (field == null)
                field = value;
        }
    }
}
```

Repro 2: RCS1213 suggests removing `dict`.

```c#
public class RCS1213Repro_2
{
    private readonly Dictionary<int, string> dict = new Dictionary<int, string> {
        [1] = "one",
        [2] = "two"
    };

    public string this[int index]
    {
        get => dict[index];
    }
}
```

Repro 3: RCS1213 suggests removing `Method()`.

```c#
public class RCS1213Repro_3
{
    private int Method() { return 0; }

    public int Zero
    {
        get
        {
            return Method();
        }
    }
}
```
Repro 4: RCS1213 suggests removing `DoSomething(int)`.

```c#
public class RCS1213Repro_4
{
    private void DoSomething(int value)
    {
        Console.WriteLine("new value = {0}", value);
    }

    public int Property
    {
        set
        {
            DoSomething(value);
        }
    }
}
```

Repro 5: RCS1213 suggests removing `Year`.

```c#
public class RCS1213Repro_5
{
    private int Year => DateTime.Now.Year;

    public int NextYear
    {
        get => Year + 1;
    }
}
```

**Comment by jhinder** [comment]
I found two more cases with false positives:

`Main` methods:

```c#
public static class RCS1213Repro_6
{
    private static void Main()
    {
    }
}
```

`nameof` operators:

```c#
public class RCS1213_Repro7
{
    public void CallWithReflection()
    {
        GetType()
            .GetMethod(nameof(PrivateMethod))
            .Invoke(this, null);
    }

    private void PrivateMethod() { }
}
```

**Comment by JosefPihrt** [comment]
https://github.com/dotnet/roslyn/issues/23902
